### PR TITLE
Support interruptible sections inside .uninterruptible

### PR DESCRIPTION
**Pull request by mvv**
A solution for #574 that adds the ability to have interruptible blocks of code inside of`.uninterruptible`.

If you look at ZIO's implementation of `bracket`
```scala
final def bracket[E, A, B](
    acquire: IO[E, A]
  )(release: A => IO[Nothing, _])(use: A => IO[E, B]): IO[E, B] =
    Ref.make[IO[Nothing, Any]](IO.unit).flatMap { m =>
      (for {
        a <- acquire.flatMap(a => m.set(release(a)).const(a)).uninterruptible
        b <- use(a)
      } yield b).ensuring(IO.flatten(m.get))
}
```
you will see that `.uninterruptible` is applied too conservatively, to both `flatMap`, which we absolutely need for the releasing logic to work, and to `acquire`, which might actually contain some interruptible steps. This leads to `Fiber`s locked in acquiring resources no longer needed by the program.

This PR adds `.interruptible` function that
* Inside a single `.uninterruptible`, makes the action it is applied to interruptible, *but* guarantees that no interruption will happen right after action produces a result. For example, in
  ```scala
  foo.interruptible.flatMap(result => bar(result)).uninterruptible
  ```
  `foo` will be interruptible, but as soon as it produces a result (using `Point`, `SyncEffect`, `AsyncEffect`, etc), the fiber will become uninterruptible.
* Otherwise behaves as the action itself, e.g. `io.interruptible` is the same as `io` when the fiber is already interruptible, and `io.interruptible.uninterruptible.uninterruptible` is the same as `io.uninterruptible.uninterruptible`.

Unfortunately, the current `AsyncEffect` implementation contains a hidden `flatMap` that can be interrupted inside an `.interruptible` without any user control over it. I've baked in a hack that prevents such interrupts from happening when `async0`'s callback returns a `Point` (e.g. in `asyncInterrupt`), but it is not a good solution. I would propose changing `async*` functions callback signatures to work with `Either` instead of `IO`. The old behavior could be when recovered, but with actual user control over what happens with that `flatMap` vs interruptibility:

```scala
object Async {
  def now[E, A](result: Either[A, E]): Async[E, A]
}
def async0[A, E](register: (Either[A, E] => Unit) => Async[E, A]): IO[E, A]
def asyncReturningIO =
  async0(k => ...)[.interruptible].flatMap { io =>
    io[.interruptible]
  }[.uninterruptible]
```

### CODE REVIEW

**Comment by mvv** [comment]
Wow, `Promise.done` is unsafe to interrupt! Will submit a PR with just that fix.

File: core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala [hunk 1]
```diff
@@ -188,8 +188,15 @@ private[zio] final class FiberContext[E, A](
                   // Enter suspended state:
                   curIo = if (enterAsync()) {
                     io.register(resumeAsync) match {
-                      case Some(io) => if (exitAsync()) io else null
-                      case None     => null
+                      case Some(io) if exitAsync() =>
+                        if (io.tag == ZIO.Tags.Succeed) {
```
Comment (mvv) [inline core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala]: The hack

File: core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala [hunk 2]
```diff
@@ -328,7 +340,17 @@ private[zio] final class FiberContext[E, A](
    * @param value The value produced by the asynchronous computation.
    */
   private[this] final val resumeAsync: IO[E, Any] => Unit =
-    io => if (exitAsync()) evaluateLater(io)
+    io =>
+      if (exitAsync()) {
+        if (io.tag == ZIO.Tags.Succeed) {
```
Comment (mvv) [inline core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala]: Also here

**Comment by mvv** [comment]
@jdegoes Any chance of giving this PR a thought before 1.0? I feel that it will be harder to land afterwards.

**Comment by jdegoes** [comment]
@mvv Sure. The problem is that it's not very sound.

To understand this, let's take a look at this:

> Inside a single .uninterruptible, makes the action it is applied to interruptible, but guarantees that no interruption will happen right after action produces a result.

OK, let's imagine that inside `acquisition`, it has _already acquired_ the resource, but has _not yet produced it_ (i.e. it is still "inside" acquisition). In this case, acquisition can be interrupted, without running the release; the release cannot be run, because even though the acquisition has produced it, it hasn't yet returned it. So the acquisition will be interrupted, the release will not be run, and the resource will be leaked.

There is a subset of problems that can support interruptible acquisition. `Managed` models this quite well—it's a generalization of `bracket`.

Have you looked to see if `Managed` can solve your current problem?

**Comment by mvv** [comment]
@jdegoes 
> OK, let's imagine that inside acquisition, it has already acquired the resource, but has not yet produced it (i.e. it is still "inside" acquisition).

Yes, when the monad has the "interruptible by default" semantics, developers **have** to think carefully about appropriate interruption points in the code they write (consider, for example, the bug in `Promise.done`). That's the price for the feature. The status quo is to have "special" `IO` subroutines that are not interruption-safe, but work when used with `bracket` through the brute force of `.uninterruptible`. I would argue that **every** subroutine should be interruption-safe, and the `.interruptible`/`.uninterruptible` pair is a tool to achieve that. For example, if you are writing an acquisition function that does some post-processing of the acquired resource:

```scala
def acquire(a): IO[R] = preProcessing(a).flatMap { b =>
  allocate(a, b).interruptible
    .flatMap { r =>
      postProcessing(r) 
    }
    .uniterruptible
}
```

Here we have reasoned about appropriate interruption points locally, knowing what `preProcessing`, `allocate` and `postProcessing` do. As the result, `acquire` is now safe to use in any context.

One corollary to this approach is that combinators should almost never use `.uninterruptible` on their arguments, instead they should focus on interruptibility of the code that combines arguments into the result.

`Managed` would still have its place as a delayed version of nested `bracket`, but without overuse of `.uninterruptible`. The current form of `Managed.make` is as bad as the `bracket` itself.

Also, resource allocation is not the only type of tasks that are sensitive to interruptions. We could have, for example, a multi-stage state transitions with designated interruption/rollback points.

**Comment by mvv** [comment]
Just to avoid confusion: `postProcessing` in the example above could also contain interruptible sections (guarded with `.onInterrupt(release(r))`) and points of failure (again, guarded).

**Comment by jdegoes** [comment]
@mvv 

Right now, `ZManaged` does not have great documentation. All the documentation is written in terms of `ZManaged.make`.

But the constructor can be used directly:

```scala
case class ZManaged[R, E, A](reserve: ZIO[R, E, ZManaged.Reservation[R, E, A]]) 
```

In this form, you provide an effect that, when run, produces a `Reservation`. 

The `Reservation` has an `acquire` effect, and a `release` effect.

The interesting thing about this formulation is that `acquire` can be interruptible. 

Can you try this and see if it solves this problem? We could always support this API directly in ZIO, e.g:

```scala
ZIO.reserve(reservation)(resource => 
  ...
)
```

`bracket`, after all, is not primitive in `ZIO`, it's just a helper function. Could be the same with `reserve`, above.

**Comment by jdegoes** [comment]
@mvv By the way, I am still interested in the possibility of `interruptible`. It does introduce potential for space leaks though, which is why I'd first like to explore if `reserve` can work.

**Comment by jdegoes** [comment]
Closing for now in favor of #696, but still quite open to this idea!

**Comment by mvv** [comment]
@jdegoes 
I think we got a bit sidetracked by the issue of changing `bracket` semantics, which in fact is a separate issue - you can have `.interruptible` **and** keep `bracket` fully uninterruptible in `acquire`. I think that `.interruptible` has merits that go beyond resource allocation:

Firstly, it can be used to simplify `bracket`, `bracketExit` and `effectAsyncM` implementations. For example, the current `bracket` can be rewritten as a much more readable:

```
final def bracket[R >: LowerR, E <: UpperE, A, A1 >: A, A2 >: A, B](
    acquire: ZIO[R, E, A],
    release: A => ZIO[R, Nothing, _],
    use: A => ZIO[R, E, B]
  ): ZIO[R, E, B] =
     for {
       r <- environment[R]
       b <- acquire.flatMap(a => use(a).interruptible.ensuring(release(a).provide(r))).uninterruptible
     } yield b
```

Look ma no `Ref`s!

Secondly, it can be used in situations that do not follow the resource allocation pattern. For example, I've recently found myself writing

```
def serveReqs(reqs: Queue[Req]): IO[E, Unit] = {
  def loop: IO[E, Unit] = {
    reqs.take.interruptible.flatMap {
      case Req1(...) =>
        processReq1
      case Req2(..) =>
        for {
          _ <- startProcessingReq2
          _ <- longOperation.interruptible.onInterruptOrError(abortReq2)
          _ <- finishProcessingReq2
       } yield ()
    }.uninterruptible *> loop
  }
  loop
}
```

I don't see how it can be expressed via `Managed` (surely not without sacrificing readability), it is a different pattern, but it is still interruption-sensitive and requires some combinators to control it.

**Comment by jdegoes** [comment]
@mvv You might be onto something, I need to see if the semantics permit the simpler definition of `bracket` you've sketched. Anyway, I probably won't have time until this weekend to take a longer look.

**Pull request closed without merging**